### PR TITLE
Adds a ELM namelist option to add a MPI_Barrier

### PR DESCRIPTION
**Pull request by bishtgautam**
A namelist option (`mpi_sync_nstep_freq`) is added to allow ELM to call an MPI_Barrier. 
By default, the MPI_Barrier is not called.

[BFB]

### CODE REVIEW

**Comment by bishtgautam** [comment]
@sarats, You can create a test via `./create_test  SMS.f09_f09.IELM --no-build`. Then, `cd <test-dir>` and add `
mpi_sync_nstep_freq = 10` in the `user_nl_elm`. Finally, build and submit the case. The output in the lnd.log file would be something like the following:

```
 Beginning timestep   : 0001-01-05_19:00:00
                        A MPI_Barrier is added in this timestep.
    Completed timestep: 0001-01-05_19:00:00
 Beginning timestep   : 0001-01-05_19:30:00
    Completed timestep: 0001-01-05_19:30:00
 Beginning timestep   : 0001-01-05_20:00:00
    Completed timestep: 0001-01-05_20:00:00
 Beginning timestep   : 0001-01-05_20:30:00
    Completed timestep: 0001-01-05_20:30:00
 Beginning timestep   : 0001-01-05_21:00:00
    Completed timestep: 0001-01-05_21:00:00
 Beginning timestep   : 0001-01-05_21:30:00
    Completed timestep: 0001-01-05_21:30:00
 Beginning timestep   : 0001-01-05_22:00:00
    Completed timestep: 0001-01-05_22:00:00
 Beginning timestep   : 0001-01-05_22:30:00
    Completed timestep: 0001-01-05_22:30:00
 Beginning timestep   : 0001-01-05_23:00:00
    Completed timestep: 0001-01-05_23:00:00
 Beginning timestep   : 0001-01-05_23:30:00
    Completed timestep: 0001-01-05_23:30:00
 Beginning timestep   : 0001-01-06_00:00:00
                        A MPI_Barrier is added in this timestep.
    Completed timestep: 0001-01-06_00:00:00
```

File: components/elm/bld/namelist_files/namelist_definition.xml [hunk 1]
```diff
@@ -2020,5 +2020,15 @@ explain the SZA dependence of direct irradiance.
 Toggle to turn on the internal mixing of dust-snow. 
 </entry>
 
+<!-- ========================================================================================  -->
+<!-- Namelist option for MPI syncing                                                           -->
+<!-- ========================================================================================  -->
+
+<entry id="mpi_sync_nstep_freq" type="integer(6)" category="history"
+       group="elm_inparm" value="0">
+The number of steps after which a MPI_Barrier is called. By default (-1), a MPI_Barrier will
```
Comment (sarats) [inline components/elm/bld/namelist_files/namelist_definition.xml:2029]: Where are you setting the default of `-1`?

File: components/elm/src/main/elm_varctl.F90 [hunk 2]
```diff
@@ -522,6 +522,10 @@ module elm_varctl
    character(len=256), public :: snicar_atm_type = 'default'
    logical, public :: use_dust_snow_internal_mixing = .false.
 
+   !----------------------------------------------------------
+   ! MPI syncing
+   !----------------------------------------------------------
+   integer, public :: mpi_sync_nstep_freq = 0
```
Comment (sarats) [inline components/elm/src/main/elm_varctl.F90:528]: I see initialization to zero here.

File: components/elm/bld/namelist_files/namelist_definition.xml (hunk 1 above)
Comment (sarats) [reply components/elm/bld/namelist_files/namelist_definition.xml:2029]: I think the logic you have works for >0 and the default is set to zero. So probably just update the comment.

Comment (bishtgautam) [reply components/elm/bld/namelist_files/namelist_definition.xml]: I have updated the comment.

**Review by sarats — APPROVED** [review]

**Comment by peterdschwartz** [comment]
Merged to master

**Pull request merged**